### PR TITLE
Leveraging SiliconCompiler to run OpenROAD-flow-scripts builds

### PR DESCRIPTION
**Pull request by WRansohoff**
This PR demonstrates a self-contained SiliconCompiler build script for the ORFS flow. It can run all of the current `OpenROAD-flow-scripts/public_tests_all` CI designs, and it does not modify any existing files in the repository (except for a `.gitignore` entry).

It uses Python logic to set up a build flow which mirrors the Makefile's steps. All of the SiliconCompiler logic is confined to the 'flow/scripts/sc/' directory, except for the  "Makefile.py" build script.

The SiliconCompiler scripts set up a JSON manifest which contains information such as results, metrics, and config values for the design / PDK / libraries.

* Shared platform configurations are set up in the `pdks/[platform]_orflow.py` and `libs/[platform]_orflow.py` scripts.
* The `targets/[platform]_orflow.py` script provides a single setup function to configure a platform's pdk/lib values, and prepare the environment variables which the ORFS TCL scripts expect.
* The `flows/orflow.py` script defines which steps to run.
* The `tools/openroad/sc_aprt.tcl` script provides a common entry point for each step, and performs some inter-task logic such as preparing liberty files and copying/renaming results.
* The top-level Makefile.py build script parses the design's "config.mk" file, sets design-specific values, and runs the build flow.

To install SiliconCompiler in a Python virtual environment:

```
git clone https://github.com/siliconcompiler/siliconcompiler.git
cd siliconcompiler
python3 -m venv create sc_env
source sc_env/bin/activate
pip3 install -r requirements.txt
pip3 install -e .
```

To build a design from the `OpenROAD-flow-scripts/flow` directory (assuming OpenROAD/yosys/KLayout are installed):

```
export SCPATH=[/path/to/OpenROAD-flow-scripts]/flow/scripts/sc
python3 Makefile.py -DESIGN_CONFIG=designs/[platform]/[design]/config.mk
```

We have tested these scripts with KLayout v0.27.8, yosys v0.13+15, and OpenROAD v2.0-3976. We were able to build GDS files for each of the designs which are run as part of the ORFS Jenkins test suite.

Once a design is built, it can be viewed in KLayout with the appropriate layer map by running:

```
sc-show -design [design_nickname]
```

Our goal is to implement every build step in the Makefile flow, but the pre-synthesis hierarchical build steps are currently commented out in `flows/orflow.py`, instead of being selectively enabled based on the $SYNTH_HIERARCHICAL flag.

### CODE REVIEW

**Comment by maliberty** [comment]
Missing DCO - see details link

**Comment by tspyrou** [comment]
@WRansohoff can you follow the procedure on the details link of the failing DCO CI to sign your commit?
@maliberty I think the security scan failure may be a false positive. Do we still test for asap7 now that it is open source?

**Comment by WRansohoff** [comment]
Sorry, I tried to follow the procedure in the link, but I got merge errors from the 'tools/OpenROAD' submodule.

I then tried to rebase and reword the commits, but it looks like I messed that up, and I'm trying to figure out how to fix it.

**Comment by WRansohoff** [comment]
Okay, I think that I fixed my merge mistakes, and I added sign-off messages to all of our commits. It looks like the DCO check is still failing, though; is this the correct format for the signoffs?

    Signed-off-by: First Last <email@example.com>

Sorry for the confusion.

**Comment by maliberty** [comment]
I think the format is fine but it needs to be on a line by itself.  Sorry for all the bother but as a first time committer I want to get this right.

**Comment by WRansohoff** [comment]
No worries, I want to get it right too! It looks like I also needed to use the email address associated with the original `git push` commands, but the DCO check is passing now.

**Comment by tspyrou** [comment]
@WRansohoff is it possible to parse the platform makefiles like you do with the design makefile? It scares me to see the platform settings duplicated. They are not static but to get updated from time to time. Now we would need to update them in 2 places.

**Comment by WRansohoff** [comment]
I think that should be possible; the logic which fetches values from the individual designs' `config.mk` files uses a general-purpose Makefile parser.

We chose to copy the PDK / liberty setup scripts that we use with SiliconCompiler in order to simplify the initial proof-of-concept, but I don't think it would be difficult to replace those hard-coded values with ones that are read from the `platforms/[platform]/config.mk` files.

**Comment by WRansohoff** [comment]
Okay, I've updated the PR to pull platform information from the relevant config.mk files. Doing that also let us remove our pdk/lib setup scripts, which significantly shortened the diff.

**Comment by tspyrou** [comment]
@vvbandeira does this failure look real to you?

**Comment by maliberty** [comment]
> @vvbandeira does this failure look real to you?

It passed in the latest Jenkins so it looks real.

**Comment by WRansohoff** [comment]
This change does not include any modifications to existing files in the repository, besides `.gitignore`. We also are not changing the CI tests to use the sc-based build flow, only adding it as an option for manual runs.

I've tried merging our branch with the tip of main, but I don't believe that anything in this PR should affect the current CI tests or design outputs.

**Comment by WRansohoff** [comment]
It looks like the test failure was resolved after merging recent changes into our branch. Sorry for the confusion.

File: flow/Makefile.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import argparse
```
Comment (maliberty) [inline flow/Makefile.py:1]: Some explanatory comments would be helpful to make clear how this is intended to be used.  Perhaps a fuller README

Comment (WRansohoff) [reply flow/Makefile.py:1]: Good point, thanks. I'll add some comments to explain the argument parsing and expand the README in flow/scripts/sc.

**Comment by WRansohoff** [comment]
Okay, I have added some more documentation in the README and top-level comments - let me know if more detail is needed.

**Review by maliberty — APPROVED** [review]

**Pull request merged**